README.md
# Badge Factor 2 - Diplomas Add-On

This WordPress plugin is an add-on to the Badge Factor 2 WordPress plugin. It
adds the ability to generate PDF diplomas with information concerning an
issued badge.


CHANGELOG.md
# Versions

## [Unreleased]

# 1.4.0

- Added: alternatively get hours from badge page if no course or no course hours

## 1.3.2

- Security update.

## 1.3.1

- Fixed: workaround for issue with accented character in PDF generation.

## 1.3.0
- Refactored: fixed performance issues.

## 1.2.2

- Fixed: added a check if $field_settings['color'] is set in Certificates_Public::generate_pdf_text()

## 1.2.1

- Fixed: added a check on $badgepage in Certificate_Controller::single()

## 1.2.0

- Added: CHANGELOG
- Fixed: php 8.0 deprecation
- Added: allow color configuration of text
- Added: use multicell for text wrapping

## 1.1.0

Initial release.


bf2-certificates.php
<?php
/**
 * Badge Factor 2 - Diplomas Addon
 * Copyright (C) 2021 ctrlweb
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * @package Badge_Factor_2_Certificates
 *
 * Plugin Name: Badge Factor 2 - Diplomas Addon
 * Plugin URI: https://github.com/ctrlwebinc/bf2-certificates
 * GitHub Plugin URI: https://github.com/ctrlwebinc/bf2-certificates
 * Description: Adds Diplomas to Badge Factor 2.
 * Author: ctrlweb
 * Version: 1.0.0
 * Author URI: https://badgefactor2.com/
 * License: GNU AGPL
 * Text Domain: bf2-certificates
 * Domain Path: /languages
 *
 * @phpcs:disable WordPress.WP.I18n.NonSingularStringLiteralDomain
 */

use BadgeFactor2\BF2_Certificates;

defined( 'ABSPATH' ) || exit;

load_plugin_textdomain( 'bf2-certificates', false, basename( dirname( __FILE__ ) ) . '/languages' );

// Define BF2_CERTIFICATES_FILE.
if ( ! defined( 'BF2_CERTIFICATES_FILE' ) ) {
	define( 'BF2_CERTIFICATES_FILE', __FILE__ );
}

// Define BF2_CERTIFICATES_ABSPATH.
if ( ! defined( 'BF2_CERTIFICATES_ABSPATH' ) ) {
	define( 'BF2_CERTIFICATES_ABSPATH', dirname( BF2_CERTIFICATES_FILE ) );
}

// Define BF2_CERTIFICATES_BASEURL.
if ( ! defined( 'BF2_CERTIFICATES_BASEURL' ) ) {
	define( 'BF2_CERTIFICATES_BASEURL', plugin_dir_url( BF2_CERTIFICATES_FILE ) );
}

// Deactivate if BadgeFactor2 is not active.
if ( ! class_exists( 'BadgeFactor2\BadgeFactor2' ) || ! \BadgeFactor2\BadgeFactor2::is_initialized() ) {
	deactivate_plugins( plugin_dir_path( __FILE__ ) . '/bf2-certificates.php' );
	die( __( 'This plugin requires Badge Factor 2.', 'bf2-certificates' ) );
}

// Include the main BF2 Certificates class.
if ( ! class_exists( 'BF2_Certificates' ) ) {
	require_once dirname( __FILE__ ) . '/src/class-bf2-certificates.php';
}

/**
 * Returns the main instance of BadgeFactor2 Certificates Add-On.
 *
 * @since  2.0.0-alpha
 * @return BF2_Certificates
 */
function bf2_certificates() {
	return BF2_Certificates::instance();
}

// Global for backwards compatibility.
$GLOBALS['badgefactor2']->certificates = bf2_certificates();


src/class-bf2-certificates.php
<?php
/**
 * Badge Factor 2
 * Copyright (C) 2021 ctrlweb
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * @package Badge_Factor_2_Certificates
 *
 * @phpcs:disable WordPress.NamingConventions.PrefixAllGlobals.VariableConstantNameFound
 */

namespace BadgeFactor2;

use BadgeFactor2\Admin\CMB2_Fields\PDF_Field;
use BadgeFactor2\Helpers\Constant;
use BadgeFactor2\Post_Types\Course_Certificate_Extension;

class BF2_Certificates {

	/**
	 * The single instance of the class.
	 *
	 * @var BF2_Certificates
	 * @since 2.0.0-alpha
	 */
	protected static $_instance = null;

	/**
	 * Main BadgeFactor 2 Certificates Add-On Instance.
	 *
	 * Ensures only one instance of BadgeFactor 2 Certificates Add-On is loaded or can be loaded.
	 *
	 * @return BF2_Certificates - Main instance.
	 * @since 1.0.0-alpha
	 * @static
	 */
	public static function instance() {
		if ( is_null( self::$_instance ) ) {
			self::$_instance = new self();
		}

		return self::$_instance;
	}

	/**
	 * BadgeFactor2 Constructor.
	 */
	public function __construct() {
		$this->define_constants();
		$this->includes();
		$this->init_hooks();
	}

	/**
	 * Badge Factor 2 Init Hooks.
	 *
	 * @return void
	 */
	public static function init_hooks() {
		//Certificates_Public::init_hooks();
		Course_Certificate_Extension::init_hooks();
		Certificates_Public::init_hooks();

		if ( is_admin() || ( defined( 'WP_CLI' ) && WP_CLI ) ) {
			Certificates_Admin::init_hooks();
			PDF_Field::init_hooks();
		}
	}

	/**
	 * Define BadgeFactor2 Constants.
	 *
	 * @return void
	 */
	private function define_constants() {
		Constant::define( 'BF2_CERTIFICATES_DATA', get_plugin_data( BF2_CERTIFICATES_FILE ) );
	}

	/**
	 * Badge Factor 2 Includes.
	 *
	 * @return void
	 */
	public function includes() {
		require_once plugin_dir_path( __FILE__ ) . '../vendor/autoload.php';
		require_once plugin_dir_path( __FILE__ ) . 'controllers/class-certificate-controller.php';
		require_once plugin_dir_path( __FILE__ ) . 'public/class-certificates-public.php';
		require_once plugin_dir_path( __FILE__ ) . 'post-types/class-course-certificate-extension.php';
		require_once plugin_dir_path( __FILE__ ) . 'helpers/class-certificate-helper.php';

		// Admin / CLI classes.
		if ( is_admin() || ( defined( 'WP_CLI' ) && WP_CLI ) ) {
			require_once plugin_dir_path( __FILE__ ) . 'admin/class-certificates-admin.php';
			require_once plugin_dir_path( __FILE__ ) . 'admin/cmb2-fields/class-pdf-field.php';
		}
	}

}


src/helpers/class-certificate-helper.php
<?php
/**
 * Badge Factor 2
 * Copyright (C) 2021 ctrlweb
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * @package Badge_Factor_2
 *
 * @phpcs:disable WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedClassFound 
 * @phpcs:disable WordPress.NamingConventions.PrefixAllGlobals.VariableConstantNameFound
 */

namespace BadgeFactor2\Helpers;

use BadgeFactor2\Certificates_Public;
use BadgeFactor2\Models\Assertion;
use BadgeFactor2\Post_Types\BadgePage;
use BadgeFactor2\Post_Types\Course;

/**
 * Basic certificate helper class.
 */
class CerficateHelper {

	/**
	 * Generates and saves certificate pdf file in disk.
	 *
     * @param string $badge_page_slug
	 * @return string
	 */
	public static function generate_and_save_certificate( $badge_page_slug ) {
		$user       = wp_get_current_user();
        $assertions = Assertion::all_for_user( $user );
        $course     = Course::get_by_badge_slug( $badge_page_slug );
        $filename = '';

        foreach ( $assertions as $a ) {
            $badgepage = BadgePage::get_by_badgeclass_id( $a->badgeclass );
            if ( false !== $badgepage ) {
                if ( $badge_page_slug === $badgepage->post_name ) {
                    $filename = Certificates_Public::generate( $course, $a, true );
                }
            }
        }
		
        return $filename;
	}

    /**
	 * generate filename based on user and badge page.
	 * 
     * @param WP_User $user
     * @param \BadgeFactor2\Post_Types\BadgePage $badge_page
	 * @return string
	 */
	public static function generate_filename( $user, $badge_page ) {
        return $user->user_login . '_' . $badge_page->post_name . '.pdf';
	}
}


src/public/class-certificates-public.php
<?php
/**
 * Badge Factor 2
 * Copyright (C) 2021 ctrlweb
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * @package Badge_Factor_2_Certificates
 *
 * @phpcs:disable WordPress.NamingConventions.PrefixAllGlobals.VariableConstantNameFound
 * @phpcs:disable WordPress.WP.I18n.NonSingularStringLiteralDomain
 */

namespace BadgeFactor2;

use BadgeFactor2\Controllers\Certificate_Controller;
use BadgeFactor2\Helpers\BuddyPress;
use BadgeFactor2\Models\BadgeClass;
use BadgeFactor2\Models\Issuer;
use setasign\Fpdi\Tcpdf\Fpdi;

class Certificates_Public {


	/**
	 * Init Hooks.
	 *
	 * @return void
	 */
	public static function init_hooks() {

		add_action( 'init', array( self::class, 'add_rewrite_tags' ), 10, 0 );
		add_action( 'init', array( self::class, 'add_rewrite_rules' ), 10, 0 );

		add_filter( 'template_include', array( Certificate_Controller::class, 'single' ) );

		add_action( 'bf2_assertion_links', array( self::class, 'certificate_link' ) );
	}


	/**
	 * Rewrite tags.
	 *
	 * @return void
	 */
	public static function add_rewrite_tags() {
		add_rewrite_tag( '%certificate%', '([^&]+)' );
	}


	public static function get_certificate_slug() {
		$options = get_option( 'bf2_certificates_settings' );
		return ! empty( $options['bf2_certificate_slug'] ) ? $options['bf2_certificate_slug'] : 'certificate';
	}

	/**
	 * Rewrite rules.
	 *
	 * @return void
	 */
	public static function add_rewrite_rules() {
		if ( BuddyPress::is_active() ) {
			// Members page managed by BuddyPress.
			$members_page = BuddyPress::get_members_page_name();

			// FIXME make certificate variable.
			$certificate_slug = self::get_certificate_slug();

			add_rewrite_rule( "{$members_page}/([^/]+)/badges/([^/]+)/{$certificate_slug}/?$", 'index.php?member=$matches[1]&badge=$matches[2]&certificate=1', 'top' );
		} else {
			// TODO Manage Members page without BuddyPress.
		}
	}


	/**
	 * Generate certificate.
	 */
	public static function generate( $course, $assertion, $badgepage = null ) {

		$plugin_data = get_plugin_data( BF2_CERTIFICATES_FILE );
		$settings    = get_option( 'bf2_certificates_settings' );

		$template_file = get_attached_file( $settings['bf2_certificate_template_id'] );

		if ( $template_file ) {
			$badge          = BadgeClass::get( $assertion->badgeclass );
			$issuer         = Issuer::get( $badge->issuer );
			$recipient      = get_user_by( 'email', $assertion->recipient->plaintextIdentity );
			$portfolio_link = bp_core_get_user_domain( $recipient->ID );

			$pdf = new Fpdi();
			$pdf->AddPage( 'L', 'Letter' );
			$pdf->setSourceFile( $template_file );

			$tpl_id = $pdf->importPage( 1 );
			$pdf->useTemplate( $tpl_id, 0, 0, null, null, true );

			foreach ( $settings as $id => $field_settings ) {

				if ( false === strpos( $id, 'bf2_certificate_template' ) && is_array( $field_settings ) ) {

					$field_settings['link'] = null;

					if ( strpos( $field_settings['text'], '$badge$' ) !== false ) {
						// $badge$
						$image = str_replace( '$badge$', $assertion->image, $field_settings['text'] );
						self::generate_pdf_image( $pdf, $field_settings, $image );
					} else {
						$text = $field_settings['text'];

						// $course$
						if ( $course && strpos( $text, '$course$' ) !== false ) {
							$text = str_replace( '$course$', $course->post_title, $text );
						} else {
							// If no course associated to badge, use badge name instead.
							$text = str_replace( '$course$', $badge->name, $text );
						}

						// $date$
						if ( strpos( $text, '$date$' ) !== false ) {
							$text = str_replace( '$date$', date( 'Y-m-d', strtotime( $assertion->issuedOn ) ), $text );
						}

						// $hours$
						if ( strpos( $text, '$hours$' ) !== false ) {
							$hours = get_post_meta( $course->ID, 'course_duration', true );
							if ( false === $hours  && null != $badgepage) {
								$hours = get_post_meta( $badgepage->ID, 'required_time_hours', true );
							}
							if ( false === $hours ) {
								$hours = 'N/A';
							}
							if ( '1' === $hours ) {
								$hours .= sprintf( ' %s', __( 'hour', $plugin_data['TextDomain'] ) );
							} else {
								$hours .= sprintf( ' %s', __( 'hours', $plugin_data['TextDomain'] ) );
							}
							$text = str_replace( '$hours$', $hours, $text );
						}

						// $issuer$
						if ( strpos( $text, '$issuer$' ) !== false ) {
							$text = str_replace( '$issuer$', $issuer->name, $text );
						}

						// $name$
						if ( strpos( $text, '$name$' ) !== false ) {
                            $str = str_replace('É','E',$recipient->display_name);
							$text = str_replace( '$name$', $str, $text );
						}

						// $portfolio$
						if ( strpos( $text, '$portfolio$' ) !== false ) {
							$text                   = str_replace( '$portfolio$', $portfolio_link, $text );
							$field_settings['link'] = $portfolio_link;
						}

						self::generate_pdf_text( $pdf, $field_settings, $text );
					}
				}
			}
			status_header( 200 );
			$pdf->Output();
			die;
		} else {
			echo 'BadgeFactor 2 settings missing!';
		}
	}


	public static function certificate_link() {
		$settings      = get_option( 'bf2_certificates_settings' );
		$template_file = get_attached_file( $settings['bf2_certificate_template_id'] );

		if ( $template_file ) {
			$plugin_data = get_plugin_data( BF2_CERTIFICATES_FILE );

			$certificate_slug = self::get_certificate_slug();

			echo sprintf( '<a target="_blank" href="%s">%s</a>', $certificate_slug, __( 'View certificate', $plugin_data['TextDomain'] ) );
		} else {
			if ( current_user_can( 'manage_badgr' ) ) {
				echo sprintf( '<a href="%s">%s</a>', admin_url() . 'admin.php?page=bf2_certificates_settings', __( 'Missing Certificate Template in settings!', $plugin_data['TextDomain'] ) );
			}
		}
	}


	/**
	 * Generate PDF Image helper function.
	 *
	 * @param Fpdi $pdf Fpdi Class.
	 * @param array $field_settings Field Settings.
	 * @param string $content Field Content.
	 * @return void
	 */
	private static function generate_pdf_image( Fpdi $pdf, array $field_settings, $content ) {
		switch ( $field_settings['align'] ) {
			case 'C':
				$pos_x = (int) $field_settings['pos_x'] - (int) round( $field_settings['width'] / 2 );
				break;
			case 'R':
				$pos_x = (int) $field_settings['pos_x'] - (int) $field_settings['width'];
				break;
			case 'L':
			default:
				$pos_x = (int) $field_settings['pos_x'];
				break;
		}
		$pdf->image( $content, $pos_x, $field_settings['pos_y'], $field_settings['width'] );
	}


	/**
	 * Generate PDF Text helper function.
	 *
	 * @param Fpdi $pdf Fpdi Class.
	 * @param array $field_settings Field Settings.
	 * @param string $content Field Content.
	 * @return void
	 */
	private static function generate_pdf_text( Fpdi $pdf, array $field_settings, $content ) {
		$arr_rgb = [0, 0, 0];
		// $pdf->SetXY( $field_settings['pos_x'], $field_settings['pos_y'] );
		$pdf->setFont( $field_settings['font'], $field_settings['style'] );
		$pdf->setFontSize( $field_settings['size'] );
		
		if ( isset( $field_settings['color'] ) && trim( $field_settings['color'] ) != '' ) {
			list( $r, $g, $b ) = array_map(
				function ( $c ) {
					return hexdec( str_pad( $c, 2, $c ) );
				},
				str_split( ltrim( $field_settings['color'], '#' ), strlen( $field_settings['color'] ) > 4 ? 2 : 1 )
			);
			$arr_rgb = [$r, $g, $b];
		}
		$pdf->SetTextColor( $arr_rgb[0], $arr_rgb[1], $arr_rgb[2] );
		
		$pdf->MultiCell(
			$field_settings['width'], // Width.
			0, // Height.
			$content, // Text.
			1, // Border.
			$field_settings['align'], // Align.
			0,
			1,
			$field_settings['pos_x'], 
			$field_settings['pos_y'],
			true, 
			0, 
			false, 
			true, 
			0
		);
	}
}


src/controllers/class-certificate-controller.php
<?php
/**
 * Badge Factor 2
 * Copyright (C) 2021 ctrlweb
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * @package Badge_Factor_2
 *
 * @phpcs:disable WordPress.WP.I18n.NonSingularStringLiteralDomain
 */

namespace BadgeFactor2\Controllers;

use BadgeFactor2\Certificates_Public;
use BadgeFactor2\Models\Assertion;
use BadgeFactor2\Page_Controller;
use BadgeFactor2\Post_Types\BadgePage;
use BadgeFactor2\Post_Types\Course;

/**
 * Certificate Controller Class.
 */
class Certificate_Controller extends Page_Controller {

	/**
	 * Post Type.
	 *
	 * @var string
	 */
	protected static $post_type = 'certificate';


	/**
	 * Outputs single template with $fields array.
	 *
	 * @param string $default_template Default template (for filter hook).
	 * @return void
	 */
	public static function single( $default_template = null ) {
		if ( get_query_var( 'member' ) && get_query_var( 'badge' ) && get_query_var( 'certificate' ) ) {
			$user       = get_user_by( 'slug', get_query_var( 'member' ) );
			$course     = Course::get_by_badge_slug( get_query_var( 'badge' ) );
			$assertions = Assertion::all_for_user( $user );

			foreach ( $assertions as $a ) {
				$badgepage = BadgePage::get_by_badgeclass_id( $a->badgeclass );
				if ( false !== $badgepage ) {
					if ( get_query_var( 'badge' ) === $badgepage->post_name ) {
						Certificates_Public::generate( $course, $a, $badgepage );
						die;
					}
				}
			}
		}
		if ( $default_template ) {
			return $default_template;
		}
	}

}


src/post-types/class-course-certificate-extension.php
<?php
/**
 * Badge Factor 2
 * Copyright (C) 2021 ctrlweb
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * @package Badge_Factor_2
 *
 * @phpcs:disable WordPress.WP.I18n.NonSingularStringLiteralDomain
 * @phpcs:disable WordPress.WP.I18n.NonSingularStringLiteralContext
 * @phpcs:disable WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound
 */

namespace BadgeFactor2\Post_Types;

/**
 * Course post type (Certificate extension).
 */
class Course_Certificate_Extension {

	/**
	 * Init hooks.
	 *
	 * @return void
	 */
	public static function init_hooks() {
		add_action( 'cmb2_admin_init', array( self::class, 'register_cpt_metaboxes' ), 20 );
	}

	/**
	 * Custom meta boxes.
	 *
	 * @return void
	 */
	public static function register_cpt_metaboxes() {

		$plugin_data = get_plugin_data( BF2_CERTIFICATES_FILE );
		$cmb         = cmb2_get_metabox( 'course_info' );

		if ( $cmb ) {
			$cmb->add_field(
				array(
					'id'           => 'certificate_template',
					'name'         => __( 'Certificate Template', $plugin_data['TextDomain'] ),
					'type'         => 'file',
					'options'      => array(
						'url' => false,
					),
					'text'         => array(
						'add_upload_file_text' => __( 'Add Template', $plugin_data['TextDomain'] ),
					),
					'query_args'   => array(
						'type' => 'application/pdf',
					),
					'preview_size' => 'small',
				)
			);
		}

		
	}
}


src/admin/class-certificates-admin.php
<?php
/**
 * Badge Factor 2
 * Copyright (C) 2021 ctrlweb
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * @package Badge_Factor_2_Certificates
 *
 * @phpcs:disable WordPress.NamingConventions.PrefixAllGlobals.VariableConstantNameFound
 * @phpcs:disable WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
 * @phpcs:disable WordPress.WP.I18n.NonSingularStringLiteralDomain
 */

namespace BadgeFactor2;

use BadgeFactor2\Models\Assertion;
use BadgeFactor2\Models\BadgeClass;
use BadgeFactor2\Post_Types\BadgePage;
use CMB2_Field;

/**
 * Certificates Admin class.
 */
class Certificates_Admin {

	/**
	 * Init Hooks.
	 *
	 * @return void
	 */
	public static function init_hooks() {
		add_action( 'cmb2_admin_init', array( self::class, 'register_settings_metaboxes' ), 10 );
		add_action( 'admin_init', array( self::class, 'manage_pdf_preview' ), 10 );
		add_action( 'cmb2_save_field_bf2_certificate_slug', array( self::class, 'save_certificate_slug' ), 99, 3 );
		add_action( 'init', array( self::class, 'flush_certificate_slug' ), 10 );
		add_action( 'init', array( self::class, 'add_custom_role_and_capabilities' ), 11 );
	}

	/**
	 * Registers Add-On settings page.
	 */
	public static function register_settings_metaboxes() {
		$plugin_data = get_plugin_data( BF2_CERTIFICATES_FILE );

		$args = array(
			'id'           => 'bf2_diplomas_settings_page',
			'menu_title'   => __( 'Diplomas', $plugin_data['TextDomain'] ),
			'title'   => __( 'Diplomas', $plugin_data['TextDomain'] ),
			'object_types' => array( 'options-page' ),
			'option_key'   => 'bf2_certificates_settings',
			'capability'   => 'manage_diplomas',
			'position' => 20,
			'icon_url' => BF2_CERTIFICATES_BASEURL . ( 'assets/images/diploma_menu_icon.png' ),
		);

		$plugins = new_cmb2_box( $args );

		// Legend.
		$legend  = '';
		$legend .= '<div style="clear:both;">';
		$legend .= '<table>';
		$legend .= sprintf( '<thead><tr><th style="padding: 5px 0">%s</th><th style="padding: 5px 0">%s</th></tr></thead>', __( 'Variable', $plugin_data['TextDomain'] ), __( 'Description', $plugin_data['TextDomain'] ) );
		$legend .= '<tbody>';
		$legend .= sprintf( '<tr><td style="padding:0"><strong>$badge$</strong></td><td style="padding:0">%s</td></tr>', __( 'Issued badge image with link', $plugin_data['TextDomain'] ) );
		$legend .= sprintf( '<tr><td style="padding:0"><strong>$course$</strong></td><td style="padding:0">%s</td></tr>', __( 'Course name with link', $plugin_data['TextDomain'] ) );
		$legend .= sprintf( '<tr><td style="padding:0"><strong>$date$</strong></td><td style="padding:0">%s</td></tr>', __( 'Badge issue date', $plugin_data['TextDomain'] ) );
		$legend .= sprintf( '<tr><td style="padding:0"><strong>$hours$</strong></td><td style="padding:0">%s</td></tr>', __( 'Course duration', $plugin_data['TextDomain'] ) );
		$legend .= sprintf( '<tr><td style="padding:0"><strong>$issuer$</strong></td><td style="padding:0">%s</td></tr>', __( 'Badge issuer name with link', $plugin_data['TextDomain'] ) );
		$legend .= sprintf( '<tr><td style="padding:0"><strong>$name$</strong></td><td style="padding:0">%s</td></tr>', __( 'Recipient name', $plugin_data['TextDomain'] ) );
		$legend .= sprintf( '<tr><td style="padding:0"><strong>$portfolio$</strong></td><td style="padding:0">%s</td></tr>', __( 'Recipient badge portfolio link', $plugin_data['TextDomain'] ) );
		$legend .= sprintf( '<tr><td style="padding:0; padding-top: 1em;"><a target="_blank" href="%s">%s</a></td></tr>', '?page=bf2_certificates_settings&preview=1', __( 'Preview', $plugin_data['TextDomain'] ) );
		$legend .= '</tbody>';
		$legend .= '</table>';
		$legend .= '</div>';

		$plugins->add_field(
			array(
				'name' => __( 'Available variables', $plugin_data['TextDomain'] ),
				'desc' => $legend,
				'type' => 'title',
				'id'   => 'bf2_certificate_instructions',
			)
		);

		$plugins->add_field(
			array(
				'name'    => __( 'Diplomas slug', BF2_DATA['TextDomain'] ),
				'desc'    => __( 'When you modify this, you need to flush your rewrite rules.', BF2_DATA['TextDomain'] ),
				'id'      => 'bf2_certificate_slug',
				'type'    => 'text',
				'default' => 'certificate',
			)
		);

		$plugins->add_field(
			array(
				'name'         => __( 'Diploma Template', $plugin_data['TextDomain'] ),
				'type'         => 'file',
				'id'           => 'bf2_certificate_template',
				'options'      => array(
					'url' => false,
				),
				'text'         => array(
					'add_upload_file_text' => __( 'Add Template', $plugin_data['TextDomain'] ),
				),
				'query_args'   => array(
					'type' => 'application/pdf',
				),
				'preview_size' => 'small',
			)
		);

		$plugins->add_field(
			array(
				'name' => __( 'Diploma Title', $plugin_data['TextDomain'] ),
				'id'   => 'bf2_certificate_title',
				'type' => 'pdf_field',
			)
		);

		$plugins->add_field(
			array(
				'name' => __( 'Diploma Subtitle', $plugin_data['TextDomain'] ),
				'id'   => 'bf2_certificate_subtitle',
				'type' => 'pdf_field',
			)
		);

		$plugins->add_field(
			array(
				'name'    => __( 'Issued Badge Image', $plugin_data['TextDomain'] ),
				'id'      => 'bf2_certificate_badge',
				'type'    => 'pdf_field',
				'default' => '$badge$',
			)
		);

		$plugins->add_field(
			array(
				'name'    => __( 'Course Name', $plugin_data['TextDomain'] ),
				'id'      => 'bf2_certificate_course',
				'type'    => 'pdf_field',
				'default' => '$course$',
			)
		);

		$plugins->add_field(
			array(
				'name'    => __( 'Recipient Text', $plugin_data['TextDomain'] ),
				'id'      => 'bf2_certificate_recipient',
				'type'    => 'pdf_field',
				'default' => 'Issued to $name$',
			)
		);

		$plugins->add_field(
			array(
				'name'    => __( 'Issued Date Text', $plugin_data['TextDomain'] ),
				'id'      => 'bf2_certificate_date',
				'type'    => 'pdf_field',
				'default' => 'Issued on $date$',
			)
		);

		$plugins->add_field(
			array(
				'name'    => __( 'Portfolio Link', $plugin_data['TextDomain'] ),
				'id'      => 'bf2_certificate_portfolio',
				'type'    => 'pdf_field',
				'default' => 'Link to recipient portfolio: $portfolio$',
			)
		);

	}


	/**
	 * Hook called on certificate_slug field save.
	 *
	 * @param boolean $updated Updated.
	 * @param string $action Action.
	 * @param CMB2_Field $instance Field instance.
	 * @return void
	 */
	public static function save_certificate_slug( bool $updated, string $action, CMB2_Field $instance ) {
		set_transient( 'flush_certificate_slug', true );
	}


	/**
	 * Hook called to verify if rewrite rules flush is required.
	 *
	 * @return void
	 */
	public static function flush_certificate_slug() {
		if ( delete_transient( 'flush_certificate_slug' ) ) {
			flush_rewrite_rules();
		}
	}


	/**
	 * Manage PDF preview in Certificates admin page.
	 *
	 * @return void
	 */
	public static function manage_pdf_preview() {
		global $pagenow;
		if ( 'admin.php' === $pagenow &&
			'bf2_certificates_settings' === $_GET['page'] &&
			isset( $_GET['preview'] ) && 
			'1' === $_GET['preview']
			) {
				self::generate_preview();
				die;
		}
	}


	/**
	 * Generates a preview.
	 *
	 * @return void
	 */
	private static function generate_preview() {
		$assertion = Assertion::random();
		$entity_id = $assertion->entityId;
		$assertion = Assertion::get( $entity_id );
		$badge     = BadgeClass::get( $assertion->badgeclass );
		$badgepage = BadgePage::get_by_badgeclass_id( $badge->entityId );
		$courses   = BadgePage::get_courses( $badgepage->ID );

		Certificates_Public::generate( $courses[0], $assertion );
	}

	/**
	 * Adds custom roles and capabilities for diploma manager
	 *
	 * @return void
	 */
	public static function add_custom_role_and_capabilities() {
		$plugin_data = get_plugin_data( BF2_BASIC_CERTIFICATES_FILE );
		
		$diploma_manager = add_role(
			'diploma-manager',
			__( 'Diploma manager', $plugin_data['TextDomain'] ),
			array(
				'read' => true,
				'manage_diplomas' => true,
			)
		);
		
		$administrator = get_role( 'administrator' );
		$administrator->add_cap( 'manage_diplomas', true );
	}
}
